### PR TITLE
[telegram] Add event channels and Answer overload

### PR DESCRIPTION
**Pull request by CrazyIvan359**
This adds 4 event trigger channels that fire when messages or query callbacks are received, presenting the entire message as the payload for easier use in rules.

* `messageEvent` provides simplified message properties
* `messageRawEvent` provides the entire message object from the Telegram library
* `callbackEvent` provides simplified callback query properties
* `callbackRawEvent` provides the entire callback query object from the Telegram library

Additionally, a new overload to the `sendTelegramAnswer` action was added that allows you to provide the `callbackId` and `messageId` that would normally be retrieved by a lookup in the handler. This can be used to answer queries that were sent before openHAB restarted by saving the `callbackId` and `messageId` outside the binding, or to answer queries directly instead of via `replyId`. `callbackId` and `messageId` are provided in both callback events.

I have tested all changes locally using rules.

This is my first time building anything Java and there is one thing that was/is not working in the build that I was not able to figure out. Both `OH-INF/binding/binding.xml` and `OH-INF/thing/thing-types.xml` are empty in the built JARs, regardless of which build process I run. Is this normal for local builds? I will be checking the automated build to see if the results are the same. Any insight would be much appreciated.

---

Another question I had is about how many things should go into one PR. I am working on other changes that will add a new action ~~and modify the return values for actions to provide data that the new action would use. It won't be a breaking change, as the return values will still be falsy if the action failed and true-ish if it succeeds.~~ Seems the concept of falsy doesn't exist in a strictly typed language like Java, so changing the return type of the actions with be a breaking change.
Should I include those commits in this PR for consiceness, or make a separate PR for them?

### CODE REVIEW

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java:436]: Please cache `callbackQuery.data().split(" ", 2)` to a local variable so it isn't called twice.

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java:332]: Similar code is used in several places. Please refactor this into a function and call that instead.

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java
Comment (CrazyIvan359) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java]: Resolved in [7fe68eb](https://github.com/openhab/openhab-addons/pull/9251/commits/7fe68eb54b09bc6a88bc1fd434b335a63ac1ae82)

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java
Comment (CrazyIvan359) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java]: Resolved in [7fe68eb](https://github.com/openhab/openhab-addons/pull/9251/commits/7fe68eb54b09bc6a88bc1fd434b335a63ac1ae82)

**Review by cpmeister — CHANGES_REQUESTED** [review]

**Comment by ZzetT** [comment]
I understand the need for a more generic approach otherwise the features for the Telegram binding will always be restricted by the APIs that we offer here. I think your PR is a step in the right direction. With the messageRawEvent we can now get the whole JSON file if somebody wants to use some fancy features (and we have many questions here regarding several features)

The coolest thing would be if we could offer the whole Telegram library API directly to the user to make use of the advanced features. But I guess this is still not possible, but could be possible if the Java API is used for writing rules!?

Otherwise, the next step would be to find a way for the user to not only have a messageRawEvent for messages received by the user but also to have a raw message that can be send to the user.

**Comment by CrazyIvan359** [comment]
> The coolest thing would be if we could offer the whole Telegram library API directly to the user to make use of the advanced features. But I guess this is still not possible, but could be possible if the Java API is used for writing rules!?

I don't believe that is possible with a binding because it isn't possible to send anything other than a string payload, but I haven't tested this. I'm also not sure it is needed in the context of home automation. This PR will suffice on the receiving side I think.

---

> but also to have a raw message that can be send to the user.

An interesting idea... I have some additional work that adds a deleteMessage action, but I have not received input on whether I should make a separate PR or just add commits to this one yet. Would a raw send not get pretty complicated, as there are different builders for different message types or am I remembering wrong?

---

I have some review changes to make, I'll probably get to them tomorrow. Silly things I didn't change before submitting because I am running a stripped down version without the state channels. Anyway, if you have suggestions let me know and I can add them at the same time.

**Comment by CrazyIvan359** [comment]
> The coolest thing would be if we could offer the whole Telegram library API directly to the user to make use of the advanced features. But I guess this is still not possible, but could be possible if the Java API is used for writing rules!?

I don't believe that is possible with a binding because it isn't possible to send anything other than a string payload, but I haven't tested this. I'm also not sure it is needed in the context of home automation. This PR will suffice on the receiving side I think.

---

> but also to have a raw message that can be send to the user.

An interesting idea... I have some additional work that adds a deleteMessage action, but I have not received input on whether I should make a separate PR or just add commits to this one yet. Would a raw send not get pretty complicated, as there are different builders for different message types or am I remembering wrong?

---

I have some review changes to make, I'll probably get to them tomorrow. Silly things I didn't change before submitting because I am running a stripped down version without the state channels. Anyway, if you have suggestions let me know and I can add them at the same time.

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java
Comment (CrazyIvan359) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java:402]: I was using `String.join()` because it is possible that `lastName` is `null` and for bots `firstName` will also be `null`.

The original code in the binding seems clunky and more (process time) costly than `String.join()`:

```java
updateChannel(LASTMESSAGENAME, (lastMessageFirstName != null || lastMessageLastName != null)
        ? new StringType((lastMessageFirstName != null ? lastMessageFirstName + " " : "")
                + (lastMessageLastName != null ? lastMessageLastName : ""))
        : UnDefType.NULL);
```

I will almalgamte both into a single operation, but what do you suggest between these two:

```java
lastMessageName = (lastMessageFirstName != null || lastMessageLastName != null)
        ? (lastMessageFirstName != null ? lastMessageFirstName : "")
        + (lastMessageLastName != null ? (lastMessageFirstName != null ? " " : "")
        + lastMessageLastName : "") : null;
```

```java
lastMessageName = String.join(" ",
        new String[] { callbackQuery.from().firstName(), callbackQuery.from().lastName() })
```

Comment (CrazyIvan359) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java:402]: Oh, maybe the second version will throw errors if one of the values is null? I have not tested that specifically and don't know enough about Java to know how it will react.

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java
Comment (CrazyIvan359) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java:403]: `chatId` is not populated for callbacks, only regular messages. I have made this change for regular messages ~80 lines above this.

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java
Comment (cpmeister) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java:402]: If you want to exclude the nulls I would suggest using a custom utility function to do it. 
```
private static String join(String delimiter, String... elements){
    StringJoiner joiner = new StringJoiner(delimiter);
    for(String element : elements){
        if(element != null){
            joiner.add(element);
        }
    }
    return joiner.toString();
}
```

**Comment by Skinah** [comment]
@CrazyIvan359 can you post what the status is? Have you made all changes and waiting for more feedback?

**Comment by CrazyIvan359** [comment]
Well I never got any input on whether multiple unrelated changes should be made in the same PR. Plus I never got around to any of the cleanup or getting the delete option added into my official work, never getting answers to my questions and lack of time meant I never did any more work on. If you have the time feel free to finish this up, it's going to be a long time before I get back to it.

**Comment by Skinah** [comment]
Sorry for the delay in my reply and thank you for the work in your contribution. The length of time to get this merged does not mean it is not valued, everyone is busy and not many people help do the reviews. I am not an expert nor someone that can merge contributions, I only help out where I can. Your changes seem to be ok'ed by the codeowner (main maintainer) which is @ZzetT and this should be quick to get moving if you finish addressing what looks to be one more area being suggested. Once done make sure you post saying it is all done and you would like it looked at again...

If PR were processed quickly it would be best to only do 1 change in a single PR, but since this is not the case myself and others do multiple changes. However please be aware that making a lot of changes can cause a longer delay to have the changes approved, so making too many changes can work against the PR vs making multiple ones that are shorter and easier to review.

Your coding looks good so I am not able to offer much help so it is probably pretty close to a state that can be merged. Otherwise we can mark it as a WIP until you are ready to contribute more.

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java:402]: > maybe the second version will throw errors if one of the values is null?

JavaDoc of String.join says:
```
Note that if an element is null, then {@code "null"} is added.
```
The current code therefore does not work as expected.
Best if you go for @cpmeister's suggestion.

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java
Comment (kaikreuzer) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/TelegramHandler.java:470]:
```suggestion
        triggerChannel(channelName, payload);
```

File: bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [inline bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml:121]: I would suggest to mark all 4 new channel types as `advanced`, since trigger channels aren't normally meant to carry data structures in JSON. For the use case at hand I understand that this is helpful, though, so I'm fine with it.

File: bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [inline bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml:123]:
```suggestion
		<label>Message Received</label>
```

File: bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [inline bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml:149]:
```suggestion
		<label>Raw Message Received</label>
```

File: bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [inline bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml:156]:
```suggestion
		<label>Query Callback Received</label>
```

File: bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [inline bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml:176]:
```suggestion
		<label>Raw Callback Query Received</label>
```

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
I'd also be fine to merge this, I only have a few minor change requests.
Sorry that @cpmeister dropped the ball on the review here, so that this PR was hanging so long.

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/action/TelegramActions.java
Comment (Skinah) [inline bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/action/TelegramActions.java:112]: Does this create a breaking change?

**Comment by Skinah** [comment]
@CrazyIvan359 Any chance we can get this sorted as I am wanting to make some changes in the binding which are in areas very close to where you have done some.

**Comment by CrazyIvan359** [comment]
@Skinah I don't have time or energy to put into this Matt, not sure when I will. If you want you can absorb the changes here into your PR or ignore my changes and whenever I have time I will update this PR with new news codebase.

None of this has to do with the delays handling this PR, I understand the scale of OH and how few people there are merging PRs.

File: bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [inline bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml:174]:
```suggestion
	<channel-type id="callbackRawEvent" advanced="true">
```

File: bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [inline bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml:154]:
```suggestion
	<channel-type id="callbackEvent" advanced="true">
```

File: bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [inline bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml:147]:
```suggestion
	<channel-type id="messageRawEvent" advanced="true">
```

File: bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [inline bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml:121]:
```suggestion
	<channel-type id="messageEvent" advanced="true">
```

Comment (Skinah) [reply bundles/org.openhab.binding.telegram/src/main/resources/OH-INF/thing/thing-types.xml]: DONE I made them all advanced.

**Comment by Skinah** [comment]
@CrazyIvan359 OK I have pushed some changes to your branch, as soon as what I suspect is an issue with the build system is fixed, I can see if it passes the build tests and ask someone to merge it. It is building fine on my system here. This way you get the credit for the hard work you have done and when you have time to work on the binding again you can look to new features and not trying to get this going. Thanks.

**Comment by Skinah** [comment]
@kaikreuzer I have applied your and others suggestions, removed some redundant code that was used and made it compile. Can you take another look to see if this can now be merged? thanks.

File: bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/action/TelegramActions.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/action/TelegramActions.java:111]: @CrazyIvan359 Can you answer this question?
It seems to be breaking the existing action. So if this is the case, please create a PR similar to https://github.com/openhab/openhab-distro/pull/1314/files for the distro.

**Review by kaikreuzer — APPROVED** [review]
lgtm, thanks!

**Comment by lolodomo** [comment]
Can someone help to describe what is the breaking change introduced by this PR ?
It needs to be listed in https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst before the final 3.2 is released.
@kaikreuzer  for information

**Comment by Skinah** [comment]
See what was changed in lines 109 to 112 of 
bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/internal/action/TelegramActions.java

I do not understand why/reasons the changes were made as this is not my PR, I tagged it as to my eyes and @kaikreuzer also asked the question for more info on this section.

**Comment by lolodomo** [comment]
Ok, I added something for that:
https://github.com/openhab/openhab-distro/pull/1352

**Comment by Skinah** [comment]
I took another look and it appears that it was overloaded with a backwards compatible action. I am pretty sure this has no breaking changes and both the old and the new will work.

**Comment by lolodomo** [comment]
> I took another look and it appears that it was overloaded with a backwards compatible action. I am pretty sure this has no breaking changes and both the old and the new will work.

Yes, I have the same feeling after looking at the code changes.
I am removing the tag and the alert.

**Pull request merged**
